cdap-app-fabric/src/main/java/io/cdap/cdap/internal/events/ProgramStatusEventPublisher.java

```diff
@@ -0,0 +1,175 @@
+                                 Iterator<ImmutablePair<String, Notification>> messages) {
+    List<ProgramStatusEvent> programStatusEvents = new ArrayList<>();
+    long publishTime = System.currentTimeMillis();
+    messages.forEachRemaining(message -> {
+      Notification notification = message.getSecond();
+      if (!notification.getNotificationType().equals(Notification.Type.PROGRAM_STATUS)) {
+        return;
+      }
+      Map<String, String> properties = notification.getProperties();
+      //get program run ID
+      String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
+      if (programStatus == null) {
+        return;
+      }
+      ProgramRunStatus programRunStatus = ProgramRunStatus.valueOf(programStatus);
+      String programRun = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      ProgramRunId programRunId = GSON.fromJson(programRun, ProgramRunId.class);
+
+      //Should event publish happen for this status
+      if (!shouldPublish(programRunStatus, programRunId)) {
+        return;
+      }
+      ProgramStatusEventDetails.Builder builder = ProgramStatusEventDetails
+        .getBuilder(programRunId.getRun(), programRunId.getApplication(), programRunId.getProgram(),
+                    programRunId.getNamespace(),
+                    programStatus,
+                    RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS));
+      String userArgsString = properties.get(ProgramOptionConstants.USER_OVERRIDES);
+      String sysArgsString = properties.get(ProgramOptionConstants.SYSTEM_OVERRIDES);
+      Type argsMapType = new TypeToken<Map<String, String>>() {
+      }.getType();
+      builder = builder
+        .withUserArgs(GSON.fromJson(userArgsString, argsMapType))
+        .withSystemArgs(GSON.fromJson(sysArgsString, argsMapType));
+      if (programRunStatus.isEndState()) {
+        builder = populateErrorDetails(builder, properties, programRunStatus);
+      }
+      ProgramStatusEventDetails programStatusEventDetails = builder.build();
+      ProgramStatusEvent programStatusEvent = new ProgramStatusEvent(publishTime, EVENT_VERSION,
+                                                                     instanceName,
+                                                                     projectName, programStatusEventDetails);
+      programStatusEvents.add(programStatusEvent);
+    });
+
+    this.eventWriters.forEach(eventWriter -> eventWriter.write(programStatusEvents));
+  }
+
+  private boolean shouldPublish(ProgramRunStatus programRunStatus, ProgramRunId programRunId) {
+    return !NamespaceId.SYSTEM.equals(programRunId.getNamespaceId());
+  }
+
+  private ProgramStatusEventDetails.Builder populateErrorDetails(
+    ProgramStatusEventDetails.Builder builder, Map<String, String> properties,
+    ProgramRunStatus status) {
+    if (status == ProgramRunStatus.KILLED) {
+      return builder;
+    }
+    if (properties.containsKey(ProgramOptionConstants.PROGRAM_ERROR)) {
```

Comment:
Just to clarify, this case is true only if the ProgramRunStatus is Failed and this key won't be present if the status is Killed?

Reply:
This key is to check if in the properties of the Notification comes the property "programError" that indicates the error because that application fails. So it will be true when the property "programError" is setted in the notification's properties.